src/schemaGenerator.js

```diff
@@ -84,6 +90,164 @@ function addResolveFunctionsToSchema(schema, resolveFunctions) {
+  constructor(len, wrappedFunction) {
+    this.len = len;
+    if (typeof wrappedFunction !== 'undefined') {
+      if (typeof wrappedFunction !== 'function') {
+        throw new Error('Second argument to MockList must be a function or undefined');
+      }
+      this.wrappedFunction = wrappedFunction;
+    }
+  }
+
+  mock(o, a, c, r, fieldType, mockTypeFunc) {
+    function randint(low, high) {
+      return Math.floor(Math.random() * (high - low + 1) + low);
+    }
+    let arr;
+    if (Array.isArray(this.len)) {
+      arr = new Array(randint(this.len[0], this.len[1]));
+    } else {
+      arr = new Array(this.len);
+    }
+    for (let i = 0; i < arr.length; i++) {
+      if (typeof this.wrappedFunction === 'function') {
+        const res = this.wrappedFunction(o, a, c, r);
+        if (res instanceof MockList) {
+          arr[i] = res.mock(o, a, c, r, getNullableType(fieldType.ofType), mockTypeFunc);
+        } else {
+          arr[i] = res;
+        }
+      } else {
+        arr[i] = mockTypeFunc(fieldType.ofType)(o, a, c, r);
+      }
+    }
+    return arr;
+  }
+}
+
+function addMockFunctionsToSchema({ schema, mocks, preserveResolvers = false } = {}) {
+  function isObject(thing) {
+    return thing === Object(thing) && !Array.isArray(thing);
+  }
+  // TODO: rewrite from using Map of mock function to using an object?
+  if (!schema) {
+    // XXX should we check that schema is an instance of GraphQLSchema?
+    throw new Error('Must provide schema to mock');
+  }
+  if (!isObject(mocks)) {
+    throw new Error('mocks must be of type Object');
+  }
+
+  // use Map internally, because that API is nicer.
+  const mockFunctionMap = new Map();
+  Object.keys(mocks).forEach((typeName) => {
+    mockFunctionMap.set(typeName, mocks[typeName]);
+  });
+
+  mockFunctionMap.forEach((mockFunction, mockTypeName) => {
+    if (typeof mockFunction !== 'function') {
+      throw new Error(`mockFunctionMap[${mockTypeName}] must be a function`);
+    }
+  });
+
+  const defaultMockMap = new Map();
+  defaultMockMap.set('Int', () => 58);
+  defaultMockMap.set('Float', () => 12.3);
+  defaultMockMap.set('String', () => 'Lorem Ipsum');
+  defaultMockMap.set('Boolean', () => false);
+  defaultMockMap.set('ID', () => '41ae7bd');
+
+  const mockType = function mockType(type, typeName, fieldName) {
+    // order of precendence for mocking:
+    // 1. if the object passed in already has fieldName, just use that
+    // --> if it's a function, that becomes your resolver
+    // --> if it's a value, the mock resolver will return that
+    // 2. if the nullableType is a list, recurse
+    // 2. if there's a mock defined for this typeName, that will be used
+    // 3. if there's no mock defined, use the default mocks for this type
+    return (o, a, c, r) => {
+      // nullability doesn't matter for the purpose of mocking.
+      const fieldType = getNullableType(type);
+
+      if (o && typeof o[fieldName] !== 'undefined') {
```

Comment:
I prefer the Map interface (wrote a comment further up in the code).
I should probably write a comment about what o, a, c and r are, but I feel
like those are actually pretty clear because they're the four arguments you
get in a resolver: the object, the arguments, the context, and the root
value (well, it's not technically the root value, but I don't know what
else to call it, because context is already taken). I feel like we might
even want to standardize these variable names. If one letter is too short,
we can use three, but I think brevity is good here, because otherwise the
lines get super long or broken, in which case the code is actually harder
to read.

On Thu, Apr 7, 2016 at 10:24 PM, Sashko Stubailo <notifications@github.com>
wrote:

> In src/schemaGenerator.js
> <https://github.com/apollostack/apollo-server/pull/14#discussion_r58981489>
> :
>
> > +  defaultMockMap.set('Boolean', () => false);
> > +  defaultMockMap.set('ID', () => '41ae7bd');
> > +
> > +  const mockType = function mockType(type, typeName, fieldName) {
> > +    // order of precendence for mocking:
> > +    // 1. if the object passed in already has fieldName, just use that
> > +    // --> if it's a function, that becomes your resolver
> > +    // --> if it's a value, the mock resolver will return that
> > +    // 2. if the nullableType is a list, recurse
> > +    // 2. if there's a mock defined for this typeName, that will be used
> > +    // 3. if there's no mock defined, use the default mocks for this type
> > +    return (o, a, c, r) => {
> > +      // nullability doesn't matter for the purpose of mocking.
> > +      const fieldType = getNullableType(type);
> > +
> > +      if (o && typeof o[fieldName] !== 'undefined') {
>
> IMO o isn't a great variable name, unless it's basically a placeholder
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apollostack/apollo-server/pull/14/files/fac3a7005782e23ba16128944fba4c2f5e924e76#r58981489>
>